src/models/room.ts

```diff
@@ -904,6 +929,71 @@ export class Room extends TypedEventEmitter<EmittedEvents, RoomEventHandlerMap>
+     * valid marker and can check the needs refresh status via
+     * `room.getTimelineNeedsRefresh()`.
+     */
+    public async refreshLiveTimeline(): Promise<void> {
+        const liveTimelineBefore = this.getLiveTimeline();
+        const forwardPaginationToken = liveTimelineBefore.getPaginationToken(EventTimeline.FORWARDS);
+        const eventsBefore = liveTimelineBefore.getEvents();
+        const mostRecentEventInTimeline = eventsBefore[eventsBefore.length - 1];
+        logger.log(
+            `[refreshLiveTimeline] for room ${this.roomId} at ` +
+            `mostRecentEventInTimeline=${mostRecentEventInTimeline.getId()} ` +
+            `forwardPaginationToken+${forwardPaginationToken}`,
+        );
+
+        // Empty out all of `this.timelineSets` but still keeps the same
+        // `timelineSet` references around so the React code updates properly
+        // and doesn't ignore the room events we emit because it checks that the
+        // `timelineSet` references are the same. We need the `timelineSet`
+        // empty so that the `client.getEventTimeline(...)` call later, will
+        // call `/context` and create a new timeline instead of returning the
+        // same one.
+        this.resetLiveTimeline(null, null);
+
+        // Get a reference to the emptied out `timelineSet`
+        //
+        // TODO: Do we want to use `this.getLiveTimeline().getTimelineSet()`
+        // instead? I think it's the same thing but what's more right?
+        const timelineSet = this.getUnfilteredTimelineSet();
+
+        // Use `client.getEventTimeline(...)` to construct a new timeline from a
+        // `/context` response state and events for the most recent event before
+        // we reset everything. The `timelineSet` we pass in needs to be empty
+        // in order for this function to call `/context` and generate a new
+        // timeline.
+        const newTimeline = await this.client.getEventTimeline(timelineSet, mostRecentEventInTimeline.getId());
+        // Set the pagination token back to the live sync token instead of using
+        // the `/context` historical token so that it matches the next response
+        // from `/sync` and we can properly continue the timeline.
+        newTimeline.setPaginationToken(forwardPaginationToken, EventTimeline.FORWARDS);
+
+        // Set our new fresh timeline as the live timeline to continue syncing
+        // forwards and back paginating from.
+        timelineSet.setLiveTimeline(newTimeline);
+        // Fixup `this.oldstate` so that `scrollback` has the pagination tokens
+        // available
+        this.fixUpLegacyTimelineFields();
+
+        // The timeline has now been refreshed ✅
+        this.setTimelineNeedsRefresh(false);
+
+        // Emit an event which clients can react to and re-load the timeline
+        // from the SDK
+        this.emit(RoomEvent.TimelineRefresh, this, timelineSet);
+    }
```

Comment:
What's a decent way to test this? I could write something in the room specs but it would just be asserting a bunch of internal structures which guarantees nothing to how it will behave in a client like Element. (Like checking that the `timelineSet` only holds the new timeline we constructed, has the correct forward pagination token, set as the live timeline, `room.timelineNeedsRefresh = true`, and that it emits events).

Could this be covered by an e2e test in another repo? (see https://github.com/matrix-org/matrix-react-sdk/pull/8354#discussion_r852588155)

Reply:
@t3chguy Any thoughts here to move this forward?

Reply:
Yeah I think for something with quite a few moving pieces, would be great to have an e2e test for it simply due to how easy to break it could be

Reply:
Added Cypress e2e tests in https://github.com/matrix-org/matrix-react-sdk/pull/8354